File: jass-to-ts/src/main/java/org/vsx/jass/JassLexer.java
package org.vsx.jass;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class JassLexer {
    /** номер строки в исходном коде */
    int line;
    /** номер символа в строке в исходном коде */
    int pos;
    /** текст исходного кода */
    String source;
    /** глобальная позиция символа в исходном коде */
    int i;

    /** алфавит операторов */
    final String opers = "=+-*/!><,";
    /** алфавит белых символов */
    final String whiteChar = " \t\r";
    /** алфавит строк */
    final String strChar = "\"";// "\"'";
    /** алфавит скобок */
    final String brac = "()[]";

    /** справочник ключевых слов */
    final HashMap<String,String> keywords = new HashMap<String,String>(Map.ofEntries(
        // Базовые типы
        Map.entry("integer", TokenKind.btyp),
        Map.entry("real",    TokenKind.btyp),
        Map.entry("boolean", TokenKind.btyp),
        Map.entry("String",  TokenKind.btyp),
        Map.entry("handle",  TokenKind.btyp),
        Map.entry("code",    TokenKind.btyp),
        // 
        Map.entry("nothing", TokenKind.btyp),
        // 
        Map.entry("type",        TokenKind.kwd),
        Map.entry("extends",     TokenKind.kwd),
        Map.entry("globals",     TokenKind.kwd),
        Map.entry("endglobals",  TokenKind.kwd),
        Map.entry("constant",    TokenKind.kwd),
        Map.entry("native",      TokenKind.kwd),
        Map.entry("takes",       TokenKind.kwd),
        Map.entry("returns",     TokenKind.kwd),
        Map.entry("function",    TokenKind.kwd),
        Map.entry("endfunction", TokenKind.kwd),
        //
        Map.entry("local", TokenKind.kwd),
        Map.entry("array", TokenKind.kwd),
        //
        Map.entry("set",      TokenKind.kwd),
        Map.entry("call",     TokenKind.kwd),
        Map.entry("if",       TokenKind.kwd),
        Map.entry("then",     TokenKind.kwd),
        Map.entry("endif",    TokenKind.kwd),
        Map.entry("else",     TokenKind.kwd),
        Map.entry("elseif",   TokenKind.kwd),
        Map.entry("loop",     TokenKind.kwd),
        Map.entry("endloop",  TokenKind.kwd),
        Map.entry("exitwhen", TokenKind.kwd),
        Map.entry("return",   TokenKind.kwd),
        Map.entry("debug",    TokenKind.kwd),
        //
        Map.entry("and", TokenKind.oper),
        Map.entry("or",  TokenKind.oper),
        //
        Map.entry("not", TokenKind.oper),
        //
        Map.entry("null",  TokenKind._null),
        Map.entry("true",  TokenKind._bool),
        Map.entry("false", TokenKind._bool)));
    
    /** режим совместимости с YDWE */
    boolean isYdweCompatible;

    ArrayList<String> operators = new ArrayList<String>(List.of("=", ",", "+", "-", "*", "/", ">", "<", "==", "!=", ">=", "<=" ));

    /** 
     * создать Лексер, без совместимости с YDWE
     */
    public JassLexer() {
        this(false);
    }

    /** 
     * создать Лексер
     * @param isYdweCompatible режим совместимости с YDWE
     */
    public JassLexer(boolean isYdweCompatible) {
        this.isYdweCompatible = isYdweCompatible;
    }

    /** 
     * Проверить на перевод строки.
     * Инкрементирует переменную line и 
     * обнуляет переменную pos,
     * если обнаружен перевод строки.
     * 
     * @return true если обнаружен перевод строки 
     */
    boolean LineBreak(boolean set) {
        if ('\n' != source.charAt(i))
            return false;
        if (set)
            return true;
        line++;
        pos = 0;
        return true;
    }

    /** 
     * Попытаться распарсить комментарий
     * @return  null если не удалось распарсить комментарий 
     */
    Token TryParseComment()
    {
        if (i == source.length() - 1) return null;
        if ('/' != source.charAt(i + 1)) return null;
        int j = i,
            l = line,
            p = pos;
        var s = source.substring(i, i + 2);
        i += 2;
        for (; i < source.length(); i++, pos++)
        {
            if ('\r' == source.charAt(i)) continue;
            if (LineBreak(true)) break;
            s += source.charAt(i);
        }

        if (i < source.length()) i--;
        return new Token(p, l, j, s, TokenKind.lcom);
    }

    /**
     * Попытаться распарсить YDWE макрос
     * @return null если не удалось распарсить макрос
     */
    Token TryParseYDWEMacro()
    {
        if (i == source.length() - 1) return null;
        int j = i,
            l = line,
            p = pos;
        var s = "//#";
        i++;
        for (; i < source.length(); i++, pos++)
        {
            if ('\r' == source.charAt(i)) continue;
            if (LineBreak(true)) break;
            s += source.charAt(i);
        }

        if (i < source.length()) i--;
        return new Token(p, l, j, s, TokenKind.ymacr);
    }

    /** 
     * Попытаться распарсить число
     * @return  null если не удалось распарсить число 
     */
    Token TryParseNumber() throws JassException
    {
        var s = "";
        int j = i,
            l = line,
            p = pos;
        boolean isOct = '0' == source.charAt(i);  // octal    0[0-7]*
        boolean isHex = '$' == source.charAt(i);  // hex      $[0-9a-fA-F]+
        boolean isXFound = false;                 //          0[xX][0-9a-fA-F]+
        boolean isDotFound = false;               // real     [0-9]+\.[0-9]*|\.[0-9]+
        boolean isNumFound = false;
        for (; i < source.length(); s += source.charAt(i), i++, pos++)
        {
            // условия при которых продолжаем
            if ('0' <= source.charAt(i) && source.charAt(i) <= '9')
            {
                if (isOct && '8' <= source.charAt(i)) throw new JassException(l, p, "wrong number: wrong octal number");
                isNumFound = true;
                continue;
            }
            // hex в формате 0[xX][0-9a-fA-F]+
            if (isOct && 1 == s.length() && ('x' == source.charAt(i) || 'X' == source.charAt(i)))
            {
                isOct = false;
                isHex = true;
                isXFound = true;
                continue;
            }
            if (isHex && (('A' <= source.charAt(i) && source.charAt(i) <= 'F') ||
                            ('a' <= source.charAt(i) && source.charAt(i) <= 'f') ||
                            (!isNumFound && '$' == source.charAt(i))))
            {
                isNumFound = true;
                continue;
            }
            if ('.' == source.charAt(i))
            {
                if (isDotFound) throw new JassException(l, p, "wrong number: multiple dot");
                if (isHex) throw new JassException(l, p, "wrong number: dot inside hex");
                if (isOct && s.length() > 1) throw new JassException(l, p, "wrong number: dot inside octadecimal");
                isOct = false;
                isDotFound = true;
                continue;
            }
            // условия при которых завершаем
            if (opers.contains(source.subSequence(i, i + 1))) break;
            if (brac.contains(source.subSequence(i, i + 1))) break;
            if (LineBreak(true)) break;
            if (whiteChar.contains(source.subSequence(i, i + 1))) break;
            // наткнулись на символ не число, не оператор, не перевод строки, не белый символ
            if (isDotFound || !isNumFound)
            {
                i = j;
                return null;
            }
            throw new JassException(l, p, "wrong number: not a number");
        }
        var typ = TokenKind.ndec;
        if (isOct) typ = TokenKind.oct;
        if (isHex) typ = isXFound ? TokenKind.xhex : TokenKind.dhex;
        if (isDotFound) typ = TokenKind.real;

        if (i < source.length()) i--;
        return new Token(p, l, j, s, typ);
    }

    /** Попытаться распарсить оператор */
    Token TryParseOperator() throws JassException
    {
        // костыль
        if (',' == source.charAt(i))
            return new Token(pos, line, i, ",", TokenKind.oper);

        var s = "";
        int j = i,
            l = line,
            p = pos;

        for (; i < source.length(); s += source.charAt(i), i++, pos++)
        {
            if (!opers.contains(source.subSequence(i, i + 1))) break;
            if (2 == s.length()) break;
        }

        if (!operators.contains(s))
        {
            s = s.substring(0, 1);
            i--;
        }

        if (i < source.length()) i--;
        return new Token(p, l, j, s, TokenKind.oper);
    }

    /** Попытаться распарсить число из 4х ASCII символов */
    Token TryParse4AsciiInt() throws JassException
    {
        var tok = TryParseString();
        if (tok.Text.length() > 6) throw new JassException(tok.Line, tok.Col, "wrong number: more than 4 ascii symbols");
        for (char c: tok.Text.toCharArray())
            if (c > '\u00ff')
                throw new JassException(tok.Line, tok.Col, "wrong number: non ascii symbol");
        // Надо проверить как оригинальный компилятор относится к 'a\'bc' последовательности
        tok.Kind = TokenKind.adec;
        return tok;
    }

    /** Попытаться распарсить строку */
    Token TryParseString() throws JassException
    {
        char eoc = source.charAt(i);

        var s = source.substring(i, i + 1);
        int j = i,
            l = line,
            p = pos;

        if (i == source.length() - 1) throw new JassException(l, p, "unclosed String");

        i++;
        for (; i < source.length(); i++, pos++)
        {
            s += source.charAt(i);
            if (source.charAt(i) == eoc && source.charAt(i - 1) != '\\') break;
            if (LineBreak(true))
                /** @warning todo: надо проверить как реагирует обычный jass */
                ;
        }

        //if (i < source.length()) i--;
        //return new Token { Col = p, Line = l, Pos = j, Text = s, Type = eoc == '"' ? TokenType.dstr : TokenType.sstr };
        return new Token(p, l, j, s, TokenKind.dstr);
    }

    /** Попытаться распарсить имя */
    Token TryParseName() throws JassException
    {
        var s = "";
        int j = i,
            l = line,
            p = pos;

        for (; i < source.length(); s += source.charAt(i), i++, pos++)
        {
            // допустимые символы
            if ('a' <= source.charAt(i) && source.charAt(i) <= 'z') continue;
            if ('A' <= source.charAt(i) && source.charAt(i) <= 'Z') continue;
            if (s.length() > 0 && '0' <= source.charAt(i) && source.charAt(i) <= '9') continue;
            if (s.length() > 0 && '_' == source.charAt(i)) continue;
            // не допустимые символы
            if (whiteChar.contains(source.subSequence(i, i + 1))) break;
            if (LineBreak(true)) break;
            if (brac.contains(source.subSequence(i, i + 1))) break;
            if (opers.contains(source.subSequence(i, i + 1))) break;
            if (strChar.contains(source.subSequence(i, i + 1))) break; // в некоторых случаях может быть норм
            // левые символы
            throw new JassException(l, p, "wrong identifier: unknown symbol");
        }
        var typ = TokenKind.name;
        if (keywords.containsKey(s)) typ = keywords.get(s);
        if ('_' == s.charAt(s.length() - 1)) throw new JassException(l, p, "wrong identifier: ends with \"_\"");

        if (i < source.length()) i--;
        return new Token(p, l, j, s, typ);
    }

    /** 
     * распарсить исходный код на токены
     * @param source исходный код на языке jass
     * @return  список из токенов
     */
    public List<Token> Tokenize(String source) throws JassException
    {
        this.source = source;
        i = 0;
        line = 0;
        pos = 0;
        var tokens = new ArrayList<Token>();

        Token tok = null;
        for (; i < source.length(); i++, pos++)
        {
            if ('/' == source.charAt(i))
            {
                // попытка распознать комментарий
                tok = TryParseComment();
                if (null != tok)
                {
                    tokens.add(tok);
                    continue;
                }
            }
            if (isYdweCompatible && '#' == source.charAt(i)) {
                tok = TryParseYDWEMacro();
                if (null != tok)
                {
                    tokens.add(tok);
                    continue;
                }
            }
            if (strChar.contains(source.subSequence(i, i + 1)))
            {
                // попытка распознать строку
                tok = TryParseString();
                if (null != tok)
                {
                    tokens.add(tok);
                    continue;
                }
            }
            // int из 4 ascii символов
            if ('\'' == source.charAt(i))
            {
                tok = TryParse4AsciiInt();
                if (null != tok)
                {
                    tokens.add(tok);
                    continue;
                }
            }
            if ('0' <= source.charAt(i) && source.charAt(i) <= '9' || '.' == source.charAt(i) || '$' == source.charAt(i))
            {
                // попытка распарсить число
                tok = TryParseNumber();
                if (null != tok)
                {
                    tokens.add(tok);
                    continue;
                }
            }
            if (opers.contains(source.subSequence(i, i + 1)))
            {
                // попытка распарсить оператор
                tok = TryParseOperator();
                if (null != tok)
                {
                    tokens.add(tok);
                    continue;
                }
            }
            if (whiteChar.contains(source.subSequence(i, i + 1)))
            {
                // игнорим пробелы
                continue;
            }
            if (brac.contains(source.subSequence(i, i + 1)))
            {
                var typ = "";
                var s = source.substring(i, i + 1);
                switch (source.charAt(i))
                {
                    case '(': typ = TokenKind.lbra; break;
                    case ')': typ = TokenKind.rbra; break;
                    case '[': typ = TokenKind.lind; break;
                    case ']': typ = TokenKind.rind; break;
                }
                tok = new Token(pos, line, i, s, typ);
                tokens.add(tok);
                continue;
            }
            if (LineBreak(false))
            {
                // записываем конец строки
                tokens.add(new Token(pos, line, i, "\n", TokenKind.ln));
                continue;
            }
            tok = TryParseName();
            tokens.add(tok);
        }
        return tokens;
    }
}

File: jass-to-ts-go/JassLexer.go
package JassToTs

import "strings"

const (
	// алфавит операторов
	opers = "=+-*/!><,"

	// алфавит белых символов
	whiteChar = " \t\r"

	// алфавит строк
	strChar = "\"" //"\"'";
	// алфавит скобок
	brac = "()[]"
)

var keywords = map[string]string{
	// Базовые типы
	"integer": btyp,
	"real":    btyp,

	"boolean": btyp,
	"string":  btyp,
	"handle":  btyp,
	"code":    btyp,
	//
	"nothing": btyp,
	//
	"type":        kwd,
	"extends":     kwd,
	"globals":     kwd,
	"endglobals":  kwd,
	"constant":    kwd,
	"native":      kwd,
	"takes":       kwd,
	"returns":     kwd,
	"function":    kwd,
	"endfunction": kwd,
	//
	"local": kwd,
	"array": kwd,
	//
	"set":      kwd,
	"call":     kwd,
	"if":       kwd,
	"then":     kwd,
	"endif":    kwd,
	"else":     kwd,
	"elseif":   kwd,
	"loop":     kwd,
	"endloop":  kwd,
	"exitwhen": kwd,
	"return":   kwd,
	"debug":    kwd,
	//
	"and": oper,
	"or":  oper,
	//
	"not": oper,
	//
	"null":  null,
	"true":  _bool,
	"false": _bool,
}

var operators = []string{"=", ",", "+", "-", "*", "/", ">", "<", "==", "!=", ">=", "<="}

type JassLexer struct {
	// номер строки в исходном коде
	line int

	// номер символа в строке в исходном коде
	pos int

	// текст исходного кода
	source string

	// глобальная позиция символа в исходном коде
	i int

	/** режим совместимости с YDWE */
	isYdweCompatible bool
}

/*
	создать Лексер
	@param isYdweCompatible режим совместимости с YDWE
*/
func NewJassLexer(isYdweCompatible bool) *JassLexer {
	return &JassLexer{isYdweCompatible: isYdweCompatible}
}

/*
	Проверить на перевод строки.
	Инкрементирует переменную this.line и
	обнуляет переменную this.pos,
	если обнаружен перевод строки.

	@return true если обнаружен перевод строки
*/
func (this *JassLexer) LineBreak(set bool) bool { // @todo make private
	if '\n' != this.source[this.i] {
		return false
	}
	if set {
		return true
	}
	this.line++
	this.pos = 0
	return true
}

// Попытаться распарсить комментарий
/// <returns>  nil если не удалось распарсить комментарий </returns>
func (this *JassLexer) TryParseComment() *Token { // @todo private
	if this.i == len(this.source)-1 {
		return nil
	}
	if '/' != this.source[this.i+1] {
		return nil
	}
	var j, l, p = this.i, this.line, this.pos
	var s = this.source[this.i : this.i+2]
	this.i += 2
	for this.i < len(this.source) {
		if '\r' == this.source[this.i] {
			continue
		}
		if this.LineBreak(true) {
			break
		}
		s += this.source[this.i:this.i]

		this.i++
		this.pos++
	}

	if this.i < len(this.source) {
		this.i--
	}
	return &Token{Col: p, Line: l, Pos: j, Text: s, Kind: lcom}
}

// Попытаться распарсить YDWE макрос
/// <returns>  nil если не удалось распарсить макрос </returns>
func (this *JassLexer) TryParseYDWEMacro() *Token { // @todo private
	if this.i == len(this.source)-1 {
		return nil
	}
	var j, l, p = this.i, this.line, this.pos
	//#warning Убрать // после добавления интерпретатора макросов
	var s = "//#"
	this.i++
	for this.i < len(this.source) {
		if '\r' == this.source[this.i] {
			continue
		}
		if this.LineBreak(true) {
			break
		}
		s += this.source[this.i:this.i]

		this.i++
		this.pos++
	}

	if this.i < len(this.source) {
		this.i--
	}
	return &Token{Col: p, Line: l, Pos: j, Text: s, Kind: ymacr}
}

// Попытаться распарсить число
/// <returns>  nil если не удалось распарсить число </returns>
func (this *JassLexer) TryParseNumber() *Token { // @todo private
	var s = ""
	var j, l, p = this.i, this.line, this.pos
	var isOct = '0' == this.source[this.i] // octal    0[0-7]*
	var isHex = '$' == this.source[this.i] // hex      $[0-9a-fA-F]+
	var isXFound = false                   //          0[xX][0-9a-fA-F]+
	var isDotFound = false                 // real     [0-9]+\.[0-9]*|\.[0-9]+
	var isNumFound = false
	for this.i < len(this.source) {
		// условия при которых продолжаем
		if '0' <= this.source[this.i] && this.source[this.i] <= '9' {
			if isOct && '8' <= this.source[this.i] {
				JassError(l, p, "wrong number: wrong octal number")
			}
			isNumFound = true
			continue
		}
		// hex в формате 0[xX][0-9a-fA-F]+
		if isOct && 1 == len(s) && ('x' == this.source[this.i] || 'X' == this.source[this.i]) {
			isOct = false
			isHex = true
			isXFound = true
			continue
		}
		if isHex && (('A' <= this.source[this.i] && this.source[this.i] <= 'F') ||
			('a' <= this.source[this.i] && this.source[this.i] <= 'f') ||
			(!isNumFound && '$' == this.source[this.i])) {
			isNumFound = true
			continue
		}
		if '.' == this.source[this.i] {
			if isDotFound {
				JassError(l, p, "wrong number: multiple dot")
			}
			if isHex {
				JassError(l, p, "wrong number: dot inside hex")
			}
			if isOct && len(s) > 1 {
				JassError(l, p, "wrong number: dot inside octadecimal")
			}
			isOct = false
			isDotFound = true
			continue
		}
		// условия при которых завершаем
		if strings.Contains(opers, this.source[this.i:this.i]) {
			break
		}
		if strings.Contains(brac, this.source[this.i:this.i]) {
			break
		}
		if this.LineBreak(true) {
			break
		}
		if strings.Contains(whiteChar, this.source[this.i:this.i]) {
			break
		}
		// наткнулись на символ не число, не оператор, не перевод строки, не белый символ
		if isDotFound || !isNumFound {
			this.i = j
			return nil
		}
		JassError(l, p, "wrong number: not a number")
	}
	var typ = ndec
	if isOct {
		typ = oct
	}
	if isHex {
		if isXFound {
			typ = xhex
		} else {
			typ = dhex
		}
	}
	if isDotFound {
		typ = real
	}

	if this.i < len(this.source) {
		this.i--
	}
	return &Token{Col: p, Line: l, Pos: j, Text: s, Kind: typ}
}

// Попытаться распарсить оператор
func (this *JassLexer) TryParseOperator() *Token { // @todo private
	// костыль
	if ',' == this.source[this.i] {
		return &Token{Col: this.pos, Line: this.line, Pos: this.i, Text: ",", Kind: oper}
	}

	var s = ""
	var j, l, p = this.i, this.line, this.pos

	for this.i < len(this.source) {
		if !strings.Contains(opers, this.source[this.i:this.i]) {
			break
		}
		if 2 == len(s) {
			break
		}
		s += this.source[this.i:this.i]
		this.i++
		this.pos++
	}

	var isContains = false
	for _, op := range operators {
		if op != s { continue }
		isContains = true
		break
	}
	if !isContains {
		s = s[0:1]
		this.i--
	}

	if this.i < len(this.source) {
		this.i--
	}
	return &Token{Col: p, Line: l, Pos: j, Text: s, Kind: oper}
}

// Попытаться распарсить число из 4х ASCII символов
func (this *JassLexer) TryParse4AsciiInt() *Token { // @todo private
	var tok = this.TryParseString()
	if len(tok.Text) > 6 {
		JassError(tok.Line, tok.Col, "wrong number: more than 4 ascii symbols")
	}
	for _, c := range tok.Text {
		if c > '\u00ff' {
			JassError(tok.Line, tok.Col, "wrong number: non ascii symbol")
		}
	}
	// Надо проверить как оригинальный компилятор относится к 'a\'bc' последовательности
	tok.Kind = adec
	return tok
}

// Попытаться распарсить строку
func (this *JassLexer) TryParseString() *Token { // @todo private
	var eoc = this.source[this.i]

	var s = this.source[this.i:this.i]
	var j, l, p = this.i, this.line, this.pos

	if this.i == len(this.source)-1 {
		JassError(l, p, "unclosed string")
	}

	this.i++
	for this.i < len(this.source) {
		s += this.source[this.i:this.i]
		if this.source[this.i] == eoc && this.source[this.i-1] != '\\' {
			break
		}
		if this.LineBreak(true) {
		}
		//#warning todo: надо проверить как реагирует обычный jass
		this.i++
		this.pos++
	}

	//if this.i < len(this.source)) this.i--;
	//return &Token { Col: p, Line: l, Pos: j, Text: s, Type = eoc == '"' ? TokenType.dstr : TokenType.sstr };
	return &Token{Col: p, Line: l, Pos: j, Text: s, Kind: dstr}
}

// Попытаться распарсить имя
func (this *JassLexer) TryParseName() *Token { // @todo private
	var s = ""
	var j, l, p = this.i, this.line, this.pos

	for this.i < len(this.source) {
		// допустимые символы
		if 'a' <= this.source[this.i] && this.source[this.i] <= 'z' {
			continue
		}
		if 'A' <= this.source[this.i] && this.source[this.i] <= 'Z' {
			continue
		}
		if len(s) > 0 && '0' <= this.source[this.i] && this.source[this.i] <= '9' {
			continue
		}
		if len(s) > 0 && '_' == this.source[this.i] {
			continue
		}
		// не допустимые символы
		if strings.Contains(whiteChar, this.source[this.i:this.i]) {
			break
		}
		if this.LineBreak(true) {
			break
		}
		if strings.Contains(brac, this.source[this.i:this.i]) {
			break
		}
		if strings.Contains(opers, this.source[this.i:this.i]) {
			break
		}
		if strings.Contains(strChar, this.source[this.i:this.i]) {
			break
		} // в некоторых случаях может быть норм
		// левые символы
		JassError(l, p, "wrong identifier: unknown symbol")

		s += this.source[this.i:this.i]
		this.i++
		this.pos++
	}
	var typ = name
	var _, ok = keywords[s];
	if ok {
		typ = keywords[s]
	}
	if '_' == s[len(s)-1] {
		JassError(l, p, "wrong identifier: ends with \"_\"")
	}

	if this.i < len(this.source) {
		this.i--
	}
	return &Token{Col: p, Line: l, Pos: j, Text: s, Kind: typ}
}

// распарсить исходный код на токены
/// <param name="source"> исходный код на языке jass</param>
/// <returns> список из токенов </returns>
func (this *JassLexer) Tokenize(source string) []*Token {
	this.source = source
	this.i = 0
	this.line = 0
	this.pos = 0
	var tokens []*Token

	var tok * Token = nil
	for this.i < len(this.source) {
		if '/' == this.source[this.i] {
			// попытка распознать комментарий
			tok = this.TryParseComment()
			if nil != tok {
				tokens = append(tokens, tok)
				continue
			}
		}
		if this.isYdweCompatible && '#' == this.source[this.i] {
			tok = this.TryParseYDWEMacro()
			if nil != tok {
				tokens = append(tokens, tok)
				continue
			}
		}
		if strings.Contains(strChar, source[this.i:this.i]) {
			// попытка распознать строку
			tok = this.TryParseString()
			if nil != tok {
				tokens = append(tokens, tok)
				continue
			}
		}
		// int из 4 ascii символов
		if '\'' == this.source[this.i] {
			tok = this.TryParse4AsciiInt()
			if nil != tok {
				tokens = append(tokens, tok)
				continue
			}
		}
		if '0' <= this.source[this.i] && this.source[this.i] <= '9' || '.' == this.source[this.i] || '$' == this.source[this.i] {
			// попытка распарсить число
			tok = this.TryParseNumber()
			if nil != tok {
				tokens = append(tokens, tok)
				continue
			}
		}
		if strings.Contains(opers, this.source[this.i:this.i]) {
			// попытка распарсить оператор
			tok = this.TryParseOperator()
			if nil != tok {
				tokens = append(tokens, tok)
				continue
			}
		}
		if strings.Contains(whiteChar, source[this.i:this.i]) {
			// игнорим пробелы
			continue
		}
		if strings.Contains(brac, source[this.i:this.i]) {
			var typ = ""
			var s = this.source[this.i:this.i]
			switch source[this.i] {
			case '(':
				typ = lbra
				break
			case ')':
				typ = rbra
				break
			case '[':
				typ = lind
				break
			case ']':
				typ = rind
				break
			}
			tok = &Token{Col: this.pos, Line: this.line, Pos: this.i, Text: s, Kind: typ}
			tokens = append(tokens, tok)
			continue
		}
		if this.LineBreak(false) {
			// записываем конец строки
			tokens = append(tokens, &Token{Col: this.pos, Line: this.line, Pos: this.i, Text: "\n", Kind: ln})
			continue
		}
		tok = this.TryParseName()
		tokens = append(tokens, tok)

		this.i++
		this.pos++
	}
	return tokens
}
